### PR TITLE
Fix cycle detection for single node workflows.

### LINKED ISSUE #3236: Workflow cycle detection fails for single step workflows
Workflows are validated by identifying flow cycles. The procedure to identify the cycles is not able to identify the cycle if there is a single step in the workflow with its output connected to its input. This is an old minor corner case issue.

### PR DESCRIPTION
**Pull request by jmchilton**
Fixes #3236.

### CODE REVIEW

File: lib/galaxy/workflow/steps.py [hunk 1]
```diff
@@ -53,7 +53,12 @@ def edgelist_for_workflow_steps( steps ):
     for step in steps:
         edges.append( ( steps_to_index[step], steps_to_index[step] ) )
         for conn in step.input_connections:
-            edges.append( ( steps_to_index[conn.output_step], steps_to_index[conn.input_step] ) )
+            output_index = steps_to_index[conn.output_step]
+            input_index = steps_to_index[conn.input_step]
+            # self connection - a cycle not detectable by topsort function.
+            if output_index == input_index:
+                raise CycleError([], 0, 0)
```
Comment (guerler) [inline lib/galaxy/workflow/steps.py:60]: Should we just raise the CycleError without keywords here? Overall it is a good fix. Thanks.

**Comment by bgruening** [comment]
@galaxybot test this

File: lib/galaxy/workflow/steps.py (hunk 1 above)
Comment (jmchilton) [reply lib/galaxy/workflow/steps.py:60]: They don't look optional to me I guess.

```
class CycleError(Exception):
    def __init__(self, sofar, numpreds, succs):
        Exception.__init__(self, "cycle in constraints",
                           sofar, numpreds, succs)
        self.preds = None
```

**Comment by bgruening** [comment]
@guerler up to you to merge this PR :)

File: lib/galaxy/workflow/steps.py (hunk 1 above)
Comment (guerler) [reply lib/galaxy/workflow/steps.py:60]: I wonder why its raised without keywords in the other two cases within this module.

**Comment by jmchilton** [comment]
Re-targeting for 17.05 - not an important fix. I'll research the error handling a bit more to answer the above question.

**Comment by guerler** [comment]
Yeah that would be good to know. It's not a big deal tho. Thanks for the fix.

**Comment by guerler** [comment]
Is this WIP?

**Comment by jmchilton** [comment]
Very slow progress - I haven't touched this since opening it. Feel free to just close it or look into changing the error handling yourself if you are eager.

**Comment by guerler** [comment]
It looks good to me and I think you are raising the error in the correct way. I will merge it when the WIP flag is taken off.

**Comment by jmchilton** [comment]
@guerler Okay - I removed the WIP label.

**Comment by guerler** [comment]
Thanks for the fix @jmchilton

**Pull request merged**